File: sdk/flutter_analytics/android/src/main/kotlin/com/myampix/analytics/MyampixAnalyticsPlugin.kt
package com.myampix.analytics

import android.content.Context
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.PurchasesUpdatedListener
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.android.installreferrer.api.InstallReferrerClient
import com.android.installreferrer.api.InstallReferrerStateListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

/**
 * MyAmpix native store-purchase autocapture (Android half).
 *
 * Wraps a Google Play [BillingClient] purely as an OBSERVER of the app's own
 * in-app-purchase transactions — the same role RevenueCat/Adjust play when
 * they hook Play Billing. This plugin never calls `launchBillingFlow`,
 * never acknowledges a purchase, and never consumes one: fulfilment stays
 * the responsibility of whatever code in the host app actually granted the
 * purchase. Forwarded payloads are consumed by
 * `lib/src/autocapture/purchase_autocapture.dart` over the
 * `myampix_analytics/purchases` [EventChannel] and re-emitted through the
 * Dart facade as the reserved `$in_app_purchase` event.
 *
 * IMPORTANT ANDROID CAVEAT (unlike iOS, where `SKPaymentQueue` is a single
 * process-wide queue any observer can attach to): Play Billing's
 * [PurchasesUpdatedListener] only fires for purchases launched through THIS
 * SAME [BillingClient] instance. If the host app purchases through its own,
 * separate `BillingClient` (e.g. RevenueCat's), this listener never
 * observes that purchase flow directly. To still surface those, this
 * plugin also calls `queryPurchasesAsync` once its client connects, which
 * returns every INAPP purchase owned by the signed-in Play account for this
 * app package — regardless of which client/session originally launched it
 * — mirroring how StoreKit replays existing transactions to a new
 * observer. Either way, this is still only ever the APP'S OWN
 * transactions: there is no API for observing another app's or another
 * user's purchases (that is all a sandboxed app can ever see).
 *
 * Kept deliberately defensive: every entry point is wrapped so a Play
 * Billing failure (missing Play Services, library mismatch, disconnects,
 * malformed callback data, ...) never crashes the host app.
 */
class MyampixAnalyticsPlugin :
    FlutterPlugin,
    PurchasesUpdatedListener {

    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var billingClient: BillingClient? = null

    /** Attribution (install-referrer) channel + its Dart listener sink. */
    private var attributionChannel: EventChannel? = null
    private var attributionSink: EventChannel.EventSink? = null
    private var referrerClient: InstallReferrerClient? = null

    /** De-dupes a purchase seen both via the live listener and the replay query. */
    private val seenTransactionIds = mutableSetOf<String>()

    /** Buffers payloads observed before Dart attaches a stream listener. */
    private val pendingPayloads = mutableListOf<Map<String, Any?>>()

    /** Buffers an install referrer fetched before Dart attaches its listener. */
    private var pendingReferrer: String? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val channel = EventChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                    eventSink = events
                    if (pendingPayloads.isNotEmpty()) {
                        pendingPayloads.forEach { events.success(it) }
                        pendingPayloads.clear()
                    }
                }

                override fun onCancel(arguments: Any?) {
                    eventSink = null
                }
            },
        )
        eventChannel = channel

        val attribution = EventChannel(binding.binaryMessenger, ATTRIBUTION_CHANNEL_NAME)
        attribution.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                    attributionSink = events
                    pendingReferrer?.let {
                        events.success(it)
                        pendingReferrer = null
                    }
                }

                override fun onCancel(arguments: Any?) {
                    attributionSink = null
                }
            },
        )
        attributionChannel = attribution

        startBillingClient(binding.applicationContext)
        maybeFetchInstallReferrer(binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        eventSink = null
        attributionChannel?.setStreamHandler(null)
        attributionChannel = null
        attributionSink = null
        try {
            billingClient?.endConnection()
        } catch (_: Throwable) {
            // Never crash the host on teardown.
        }
        billingClient = null
        try {
            referrerClient?.endConnection()
        } catch (_: Throwable) {
            // Never crash the host on teardown.
        }
        referrerClient = null
    }

    private fun startBillingClient(context: Context) {
        try {
            val client = BillingClient.newBuilder(context)
                .setListener(this)
                .enablePendingPurchases()
                .build()
            billingClient = client
            client.startConnection(
                object : BillingClientStateListener {
                    override fun onBillingSetupFinished(billingResult: BillingResult) {
                        if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                            replayExistingPurchases(client)
                        }
                    }

                    override fun onBillingServiceDisconnected() {
                        // Never crash; a future purchase update (if the
                        // client reconnects) simply resumes forwarding.
                    }
                },
            )
        } catch (_: Throwable) {
            // Defensive: Play Billing may be unavailable at runtime (no
            // Play Services, library mismatch, ...). Never crash the host.
        }
    }

    /**
     * Replays purchases that already exist for this account/app so a
     * purchase made through a DIFFERENT billing client (see class doc)
     * still gets picked up, much like StoreKit replays its queue to a
     * freshly attached observer.
     */
    private fun replayExistingPurchases(client: BillingClient) {
        try {
            val params = QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.INAPP)
                .build()
            client.queryPurchasesAsync(params) { billingResult, purchases ->
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                    purchases.forEach { handlePurchase(client, it) }
                }
            }
        } catch (_: Throwable) {
            // Best-effort only.
        }
    }

    override fun onPurchasesUpdated(billingResult: BillingResult, purchases: MutableList<Purchase>?) {
        try {
            if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) return
            val client = billingClient ?: return
            purchases?.forEach { handlePurchase(client, it) }
        } catch (_: Throwable) {
            // Never crash the host from a billing callback.
        }
    }

    private fun handlePurchase(client: BillingClient, purchase: Purchase) {
        try {
            if (purchase.purchaseState != Purchase.PurchaseState.PURCHASED) return
            val transactionId = purchase.orderId ?: purchase.purchaseToken
            if (transactionId.isNullOrEmpty()) return
            if (!seenTransactionIds.add(transactionId)) return // already forwarded

            fetchProductPricing(client, purchase.products) { priceMicros, currency ->
                if (purchase.products.isEmpty()) return@fetchProductPricing
                purchase.products.forEach { productId ->
                    emit(
                        mapOf(
                            "productId" to productId,
                            "price" to priceMicros?.let { it / 1_000_000.0 },
                            "currency" to currency,
                            "quantity" to purchase.quantity,
                            "transactionId" to transactionId,
                            "store" to "play_store",
                        ),
                    )
                }
            }
        } catch (_: Throwable) {
            // Never crash the host from a billing callback.
        }
    }

    /**
     * Best-effort price/currency lookup for the purchase's first product
     * via `queryProductDetailsAsync`. Play Billing purchases do not carry
     * price themselves — only `ProductDetails` does. Failure (or an app
     * with no matching product details, e.g. mismatched sandbox config)
     * degrades to `null`/`null`, never a crash.
     */
    private fun fetchProductPricing(
        client: BillingClient,
        productIds: List<String>,
        onResult: (priceMicros: Long?, currency: String?) -> Unit,
    ) {
        val firstProductId = productIds.firstOrNull()
        if (firstProductId == null) {
            onResult(null, null)
            return
        }
        try {
            val product = QueryProductDetailsParams.Product.newBuilder()
                .setProductId(firstProductId)
                .setProductType(BillingClient.ProductType.INAPP)
                .build()
            val params = QueryProductDetailsParams.newBuilder()
                .setProductList(listOf(product))
                .build()
            client.queryProductDetailsAsync(params) { billingResult, productDetailsList ->
                val offer = productDetailsList.firstOrNull()?.oneTimePurchaseOfferDetails
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK && offer != null) {
                    onResult(offer.priceAmountMicros, offer.priceCurrencyCode)
                } else {
                    onResult(null, null)
                }
            }
        } catch (_: Throwable) {
            onResult(null, null)
        }
    }

    private fun emit(payload: Map<String, Any?>) {
        val sink = eventSink
        if (sink != null) {
            sink.success(payload)
        } else {
            pendingPayloads.add(payload)
        }
    }

    /**
     * Fetches the Google Play install referrer ONCE (first launch) and
     * forwards its raw `utm_*`-carrying query string over the attribution
     * channel. A persisted flag guarantees the referrer is only ever
     * forwarded once, so the Dart side emits a single first-touch
     * `$campaign_touch` (`$attribution_source: "install_referrer"`) — not one
     * per launch. iOS has no install-referrer equivalent; its plugin half
     * registers this channel as a documented no-op.
     *
     * Defensive throughout: a missing Play Store, service disconnect, or
     * malformed referrer degrades to "no touch", never a crash.
     */
    private fun maybeFetchInstallReferrer(context: Context) {
        try {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            if (prefs.getBoolean(PREF_REFERRER_FORWARDED, false)) return
            val client = InstallReferrerClient.newBuilder(context).build()
            referrerClient = client
            client.startConnection(
                object : InstallReferrerStateListener {
                    override fun onInstallReferrerSetupFinished(responseCode: Int) {
                        try {
                            if (responseCode == InstallReferrerClient.InstallReferrerResponse.OK) {
                                val referrer = client.installReferrer.installReferrer
                                if (!referrer.isNullOrEmpty()) {
                                    emitReferrer(referrer)
                                    prefs.edit()
                                        .putBoolean(PREF_REFERRER_FORWARDED, true)
                                        .apply()
                                }
                            }
                        } catch (_: Throwable) {
                            // Best-effort only: never crash on a referrer read.
                        } finally {
                            try {
                                client.endConnection()
                            } catch (_: Throwable) {
                                // Never crash on teardown.
                            }
                        }
                    }

                    override fun onInstallReferrerServiceDisconnected() {
                        // Never crash; the persisted flag stays unset so a
                        // future launch can retry the one-time fetch.
                    }
                },
            )
        } catch (_: Throwable) {
            // Install Referrer may be unavailable (no Play Store, ...). Never
            // crash the host.
        }
    }

    private fun emitReferrer(referrer: String) {
        val sink = attributionSink
        if (sink != null) {
            sink.success(referrer)
        } else {
            pendingReferrer = referrer
        }
    }

    private companion object {
        const val CHANNEL_NAME = "myampix_analytics/purchases"
        const val ATTRIBUTION_CHANNEL_NAME = "myampix_analytics/attribution"
        const val PREFS_NAME = "myampix_analytics"
        const val PREF_REFERRER_FORWARDED = "install_referrer_forwarded"
    }
}

File: sdk/flutter_analytics/ios/Classes/MyampixAnalyticsPlugin.swift
import Flutter
import StoreKit
import UIKit

/// MyAmpix native store-purchase autocapture (iOS half).
///
/// Registers an `SKPaymentTransactionObserver` on the shared, process-wide
/// `SKPaymentQueue` and forwards `.purchased` transactions to
/// Dart over the `myampix_analytics/purchases` `EventChannel`. Consumed by
/// `lib/src/autocapture/purchase_autocapture.dart`, which re-emits them
/// through the Dart facade as the reserved `$in_app_purchase` event.
///
/// IMPORTANT — this is a PASSIVE observer, exactly like RevenueCat/Adjust's
/// StoreKit hooks: it never calls `finishTransaction`. Finishing/acking a
/// transaction remains the responsibility of whatever code in the host app
/// actually fulfils the purchase (the app's own StoreKit integration,
/// RevenueCat, ...). We only look; we never touch.
///
/// A sandboxed plugin can only ever see transactions that pass through this
/// process's `SKPaymentQueue` — i.e. purchases made by THIS app install.
/// There is no API for observing another app's or another user's
/// purchases; "native" here means "the app's own StoreKit transactions",
/// not store-wide telemetry.
///
/// Kept deliberately defensive: every callback is wrapped so a StoreKit
/// failure or unexpected transaction shape never crashes the host app.
public class MyampixAnalyticsPlugin: NSObject, FlutterPlugin, SKPaymentTransactionObserver, SKProductsRequestDelegate, FlutterStreamHandler {
  private static let channelName = "myampix_analytics/purchases"
  private static let attributionChannelName = "myampix_analytics/attribution"

  private var eventSink: FlutterEventSink?

  /// Transactions observed before Dart attaches a stream listener (the queue
  /// replays them at cold start, before `runApp` completes) are buffered here
  /// and flushed as soon as `onListen` runs, so nothing is silently dropped.
  /// The replay is why Dart de-dupes: this buffer is faithful, not filtered.
  private var pendingPayloads: [[String: Any?]] = []

  /// Best-effort product price/currency cache, keyed by productIdentifier.
  /// Populated asynchronously via `SKProductsRequest`; a transaction is
  /// forwarded immediately regardless of whether this cache is warm yet —
  /// price/currency are simply `nil` until/unless a lookup completes.
  private var productCache: [String: (price: Double, currency: String?)] = [:]
  private var pendingProductRequests: Set<String> = []
  private var productIdsByRequest: [ObjectIdentifier: [String]] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = MyampixAnalyticsPlugin()
    let eventChannel = FlutterEventChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(instance)
    SKPaymentQueue.default().add(instance)
    // Keeps the plugin instance alive for the lifetime of the engine so the
    // SKPaymentQueue observer registration above is never deallocated.
    registrar.publish(instance)

    // Marketing-attribution channel — iOS HAS NO INSTALL-REFERRER EQUIVALENT.
    // Unlike Android's Google Play `InstallReferrerClient` (see
    // MyampixAnalyticsPlugin.kt), Apple exposes no API for a generic
    // install-attribution string (SKAdNetwork is a privacy-preserving,
    // postback-only mechanism, not a utm_* referrer). So this half registers
    // the `myampix_analytics/attribution` channel purely to keep the Dart
    // `EventChannel` well-formed, and NEVER emits: iOS attribution is
    // deep-link-only via `MyAmpix.trackDeepLink`. This mirrors the honest
    // Play-Billing caveat documented on the purchase channel above.
    let attributionChannel = FlutterEventChannel(
      name: attributionChannelName,
      binaryMessenger: registrar.messenger()
    )
    attributionChannel.setStreamHandler(MyampixAttributionNoopStreamHandler())
  }

  deinit {
    SKPaymentQueue.default().remove(self)
  }

  // MARK: - FlutterStreamHandler

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    if !pendingPayloads.isEmpty {
      pendingPayloads.forEach { events($0) }
      pendingPayloads.removeAll()
    }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  // MARK: - SKPaymentTransactionObserver

  /// Only `.purchased` is forwarded. `.restored` is deliberately NOT: it is
  /// what `restoreCompletedTransactions()` replays after a login or a
  /// reinstall, and for an auto-renewable subscription that replay is the
  /// product's ENTIRE billing history at once. Those are not sales — the
  /// money was already reported when each renewal originally came through as
  /// `.purchased` — so counting them would inflate revenue by however many
  /// periods the subscriber has been paying.
  ///
  /// This is only half the guard. StoreKit also re-delivers `.purchased`
  /// transactions to every newly attached observer, so a plain cold start
  /// replays them too; the Dart side (`PurchaseAutocapture`) holds the
  /// persisted seen-transaction set that makes each one emit exactly once.
  public func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
    for transaction in transactions {
      switch transaction.transactionState {
      case .purchased:
        forward(transaction)
        maybeFetchProduct(for: transaction.payment.productIdentifier)
      default:
        break
      }
    }
  }

  private func forward(_ transaction: SKPaymentTransaction) {
    let productId = transaction.payment.productIdentifier
    // `transactionIdentifier` is per-transaction, so each subscription
    // renewal keeps its own id and reports its own revenue — unlike
    // `original.transactionIdentifier`, which every renewal of a
    // subscription shares and which is therefore only a last-resort
    // fallback, never the primary key.
    let transactionId = transaction.transactionIdentifier ?? transaction.original?.transactionIdentifier
    guard !productId.isEmpty, let transactionId = transactionId, !transactionId.isEmpty else {
      // Never forward a payload the Dart side cannot map onto the §4
      // contract's required properties.
      return
    }
    let cached = productCache[productId]
    let payload: [String: Any?] = [
      "productId": productId,
      "price": cached?.price,
      "currency": cached?.currency,
      "quantity": transaction.payment.quantity,
      "transactionId": transactionId,
      "store": "app_store",
    ]
    if let sink = eventSink {
      sink(payload)
    } else {
      pendingPayloads.append(payload)
    }
  }

  private func maybeFetchProduct(for productId: String) {
    guard productCache[productId] == nil, !pendingProductRequests.contains(productId) else { return }
    pendingProductRequests.insert(productId)
    let request = SKProductsRequest(productIdentifiers: [productId])
    request.delegate = self
    productIdsByRequest[ObjectIdentifier(request)] = [productId]
    request.start()
  }

  // MARK: - SKProductsRequestDelegate

  public func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
    let requestedIds = productIdsByRequest.removeValue(forKey: ObjectIdentifier(request)) ?? []
    for product in response.products {
      pendingProductRequests.remove(product.productIdentifier)
      productCache[product.productIdentifier] = (
        price: product.price.doubleValue,
        currency: product.priceLocale.currencyCode
      )
    }
    for id in requestedIds where productCache[id] == nil {
      pendingProductRequests.remove(id)
    }
  }

  public func request(_ request: SKRequest, didFailWithError error: Error) {
    // Best-effort only: leave price/currency nil for these products next
    // time they're forwarded. Never crash on a products-request failure.
    guard let productsRequest = request as? SKProductsRequest else { return }
    let requestedIds = productIdsByRequest.removeValue(forKey: ObjectIdentifier(productsRequest)) ?? []
    requestedIds.forEach { pendingProductRequests.remove($0) }
  }
}

/// No-op stream handler for the `myampix_analytics/attribution` channel on
/// iOS. iOS has no install-referrer equivalent, so this handler accepts the
/// Dart listener and never emits — iOS marketing attribution is deep-link
/// only via `MyAmpix.trackDeepLink`. See the caveat in
/// `MyampixAnalyticsPlugin.register`.
private class MyampixAttributionNoopStreamHandler: NSObject, FlutterStreamHandler {
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    // Intentionally emits nothing.
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    return nil
  }
}
